lib/appFactory.js

```diff
@@ -0,0 +1,85 @@
+const express = require('express');
+const path = require('path');
+const { loggingMiddleware } = require('./logger');
+const indexRouter = require('./routes/index');
+const signupRouter = require('./routes/signup');
+const webFingerRouter = require('./routes/webfinger');
+const oAuthRouter = require('./routes/oauth');
+const storageCommon = require('./routes/storage_common');
+const errorPage = require('./util/errorPage');
+const helmet = require('helmet');
+const shorten = require('./util/shorten');
+
+module.exports = function ({ hostIdentity, jwtSecret, account, store, basePath = '' }) {
+  if (basePath && !basePath.startsWith('/')) { basePath = '/' + basePath; }
+
+  const app = express();
+  app.locals.basePath = basePath;
+
+  // view engine setup
+  app.engine('.html', require('ejs').__express);
+  app.engine('.xml', require('ejs').__express);
+  app.set('view engine', 'html');
+  app.set('views', path.join(__dirname, 'views'));
+
+  express.static.mime.define({ 'text/javascript': ['js'] });
+
+  app.set('account', account);
+
+  app.use(loggingMiddleware);
+  app.use(helmet({
+    contentSecurityPolicy: {
+      directives: {
+        sandbox: ['allow-scripts', 'allow-forms', 'allow-popups', 'allow-same-origin'],
+        defaultSrc: ['\'self\''],
+        scriptSrc: ['\'self\''],
+        scriptSrcAttr: ['\'none\''],
+        styleSrc: ['\'self\''],
+        imgSrc: ['\'self\''],
+        fontSrc: ['\'self\''],
+        objectSrc: ['\'none\''],
+        childSrc: ['\'none\''],
+        connectSrc: ['\'none\''],
+        baseUri: ['\'self\''],
+        frameAncestors: ['\'none\''],
+        formAction: (process.env.NODE_ENV === 'production' ? ['https:'] : ['https:', 'http:']), // allows redirect to any RS app
+        upgradeInsecureRequests: []
+      }
+    }
+  }));
+  app.use(express.urlencoded({ extended: false }));
+  app.use(`${basePath}/assets`, express.static(path.join(__dirname, 'assets')));
+
+  app.use(`${basePath}/`, indexRouter);
+
+  app.use(`${basePath}/signup`, signupRouter);
+
+  app.use([`${basePath}/.well-known`, `${basePath}/webfinger`], webFingerRouter);
+
+  app.use(`${basePath}/oauth`, oAuthRouter(hostIdentity, jwtSecret));
+  app.use(`${basePath}/storage`, storageCommon(hostIdentity, jwtSecret));
+  app.use(`${basePath}/storage`, store);
+
+  // catches paths not handled and returns Not Found
+  app.use(basePath, function (req, res, next) {
+    const name = req.path.slice(1);
+    errorPage(req, res, 404, { title: 'Not Found', message: `“${name}” doesn't exist` });
+  });
```

Comment:
Nice.  It is still surprising to me how many apps I have worked on miss this simple fall through routing handler.